app/ota_client_stub.py

```diff
@@ -15,23 +20,49 @@ def __init__(self):
         self._ecu_info = EcuInfo()
         self._ota_client_call = OtaClientCall("50051")
 
-    def update(self, request):
-        response = []
+        # dispatch the requested operations to threadpool
+        self._executor = ThreadPoolExecutor()
+        # a dict to hold the future for each requests if needed
+        self._future = dict()
+    
+    def terminate(self):
+        self._executor.shutdown()
 
+    async def update(self, request):
         # secondary ecus
-        secondary_ecus = self._ecu_info.get_secondary_ecus()
+        tasks, secondary_ecus = [], self._ecu_info.get_secondary_ecus()
         for secondary in secondary_ecus:
-            entry = OtaClientStub._find_request(request.ecu, secondary)
-            if entry:
-                r = self._ota_client_call.update(request, secondary["ip_addr"])
-                response.append(r)
+            if OtaClientStub._find_request(request.ecu, secondary):
+                tasks.append(asyncio.create_task(
+                    self._ota_client_call.update(request, secondary["ip_addr"]),
+                    name = secondary, # register the task name with sub_ecu id
+                    ))
+        sub_ecu_update_aws = asyncio.gather(*tasks)
```

Comment:
My understanding is this line blocks until all sub ecu update requests are finished?
And this update is called by asyncio.run().

Reply:
Ah, I do not mention error scenarios.
My question is: Does `asyncio.gather` wait until sub-ecu update is finished?